plejd/mqtt.js

```diff
@@ -87,13 +87,15 @@ class MqttClient extends EventEmitter {
     this.client.on('connect', () => {
       logger('connected to MQTT.');
 
-      this.client.subscribe(startTopic, (err) => {
-        if (err) {
-          logger('error: unable to subscribe to ' + startTopic);
-        }
+      for (const startTopic of startTopics) {
+        this.client.subscribe(startTopic, (err) => {
+          if (err) {
+            logger('error: unable to subscribe to ' + startTopic);
+          }
 
-        self.emit('connected');
-      });
+          self.emit('connected');
+        });
+      }
 
```

Comment:
This will trigger the connected event twice, resulting in the event handler attached will be executed twice as well. Need to look into the effects of that or rewrite to only trigger once.

Reply:
Right. Good catch. I'll look into this.

Reply:
It don't think it will trigger twice, since you can never have both of them configured.

(at least if I have understood these correctly)

https://www.home-assistant.io/docs/mqtt/birth_will/

https://github.com/home-assistant/core/blob/0e068a5f394e4cddde918d7f1cd4343000e4f002/homeassistant/components/mqtt/__init__.py#L847

Reply:
Seems it can be a list, but I don't know how reasonable it is for a user to configure both...

![image](https://user-images.githubusercontent.com/449347/106035958-ff731f00-60d4-11eb-8aa2-53b92b7e8717.png)

Reply:
Wow, you are correct, missed that.
Any chance there is a bug in their documentation?

Then I guess this needs to be fixed :)

Reply:
Indeed, seems to be a bug in their documentation, tried setting it now and got an error, the validator wants a dictionary, not a list.

This config
```
mqtt:
  birth_message:
    - topic: 'hass/status'
      payload: 'online'
    - topic: 'homeassistant/status' 
      payload: 'online'
  will_message:
    - topic: 'hass/status'
      payload: 'offline'
    - topic: 'homeassistant/status' 
      payload: 'offline'
```

Gave this error:


```
Logger: homeassistant.components.hassio
Source: components/hassio/__init__.py:420
Integration: Hass.io (documentation, issues)
First occurred: 9:44:55 AM (2 occurrences)
Last logged: 9:48:29 AM

    Invalid config for [mqtt]: expected a dictionary for dictionary value @ data['mqtt']['birth_message']. Got [OrderedDict([('topic', 'hass/status'), ('payload', 'maaan')]), OrderedDict([('topic', 'homeassistant/status'), ('payload', 'maaan-boy')])] expected a dictionary for dictionary value @ data['mqtt']['will_message']. Got [OrderedDict([('topic', 'hass/status'), ('payload', 'womaaan')]), OrderedDict([('topic', 'homeassistant/status'), ('payload', 'womaaan-girl')])]. (See /config/configuration.yaml, line 6).
    Invalid config for [mqtt]: expected a dictionary for dictionary value @ data['mqtt']['birth_message']. Got [OrderedDict([('topic', 'hass/status'), ('payload', 'online')]), OrderedDict([('topic', 'homeassistant/status'), ('payload', 'online')])] expected a dictionary for dictionary value @ data['mqtt']['will_message']. Got [OrderedDict([('topic', 'hass/status'), ('payload', 'offline')]), OrderedDict([('topic', 'homeassistant/status'), ('payload', 'offline')])]. (See /config/configuration.yaml, line 6).
```

Reply:
That's not the problem with this line, though.
The problem is that this function will emit `connected` twice, making the main system think it lost connection and regained it... I think.
I don't think it hurts, but it's unnecessary.

Reply:
so - move that line to after the foreach brace basically then!

Reply:
Help me understand, I don't belive so

It will run the content of the loop two times, agreed?

First time around it will subscribe on 'hass/status' and if it ever gets a message on that topic it will run the callback you provided. Which is:

 ```
     (err) => {
          if (err) {
            logger('error: unable to subscribe to ' + startTopic);
          }

          self.emit('connected');
        }
```


Second time around it will subscribe on 'homeassistant/status' and if it ever gets a message on that topic it will run the callback you provided (same callback)

The callback runs when the topic gets a message, since you can only have one or the other, you will never run the callback twice?

Reply:
No, the callback runs when the client gets a response from the mqtt server that the subscription has been registered, not when there's a message on the topic.

Turns out it's very simple to fix, though, since `mqtt.subscribe()` can also accept a *list* of topics...